### PR TITLE
SNOW-1167007: Validate no null values for not null cols

### PR DESCRIPTION
**Pull request by sfc-gh-wtrefon**
Validate if row contains null values for non-nullable columns. If yes, add such column names to InsertValidationResponse error as _NullValueForNotNullColNames_ so it can be used for schema evolution.

### CODE REVIEW

File: src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java [hunk 1]
```diff
@@ -442,6 +442,26 @@ Set<String> verifyInputColumns(
...
     }
 
+    // Check for null values in not-nullable columns in the row
+    List<String> nullValueNotNullCols = new ArrayList<>();
+    for (String columnName : this.nonNullableFieldNames) {
+      if (inputColNamesMap.containsKey(columnName)
+          && row.get(inputColNamesMap.get(columnName)) == null) {
+        nullValueNotNullCols.add(statsMap.get(columnName).getColumnDisplayName());
+      }
+    }
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:452]: We already throw an exception at the end of `parseColumnValueToParquet` when we insert null into a non-nullable column, should be just catch that and add the col name to the nullValueNotNullCols list?

Comment (sfc-gh-wtrefon) [reply src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:452]: I'm not sure if it is the right place. The `verifyInputColumns` is called before and sounds like the right place for such validation. `verifyInputColumns` is also in the abstract class, the `parseColumnValueToParquet` only in one implementation (Parquet).

If we catched the exception in insertRows function we would only add single column to the nullValueNotNullCols even if there is more columns with null values. We would need to pass the error object into the `addRow` function to catch all the exception that may be thrown in a loop.

Therefore i think the `verifyInputColumns` is much better place for the validation

File: src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java [hunk 2]
```diff
@@ -442,6 +442,26 @@ Set<String> verifyInputColumns(
               "Values for all non-nullable columns must be specified, rowIndex:%d", rowIndex));
     }
 
+    // Check for null values in not-nullable columns in the row
+    List<String> nullValueNotNullCols = new ArrayList<>();
```
Comment (sfc-gh-lshcharbaty) [inline src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:446]: Does it make sense to differentiate between `missingCols` and `nullValueNotNullCols`?
Is there a conceptual difference between the cases when a column is present with `null` value and when it isn't present at all.

Comment (sfc-gh-wtrefon) [reply src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:446]: i was thinking about this, the only thing is the naming of missingCols. But if we do like this than we don't need to adapt anything in kafka push code. What u think @sfc-gh-tzhang ?

Comment (sfc-gh-tzhang) [reply src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:446]: I'm fine with combing it and we will need to update the name to make it clear, the only concern is that we're updating a public API, but I highly doubt any customer is using this.

File: src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java (hunk 1 above)
Comment (sfc-gh-tzhang) [reply src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:452]: I see, makes sense to me, my concern was that in the worst case, we're iterating through all the data again to check this which is not ideal so I was hoping that we can do this as part of data validation.

**Review by sfc-gh-tzhang — APPROVED** [review]
LGTM, thanks!

**Pull request merged**